doc/design/mount-config-design.md

```diff
@@ -0,0 +1,115 @@
+# Automatic filesystem mount configuration design
+
+During startup we need the ability to mount directories from the insecure host into the enclave.
+Currently an application can do this through a system call,
+but this needs to be automated at start-up if the necessary configuration is given.
+Configuration needed for a mount includes the mount source path which is on an insecure filesystem,
+mount target directory within the TEE filesystem,
+and other data relating configuration like if it is read-only or not.
+
+## Overview
+
+Currently three file-systems are supported by Mystikos:
+
+* ramfs is a CPIO archive that is totally TEE memory resident once loaded.
+It is used for rootfs file-systems and is included as part of the tee integrity measurement.
+A ramfs archive can also be included in the rootfs image and an application can mount that on a target mount point.
+No encryption is supported.
+
+* ext2fs filesystem image that is hosted in the non-TEE filesystem for a rootfs.
+It is currently used for the rootfs filesystem and can be integrity protected and encrypted.
+It can also be mounted from the rootfs directory to a target mount point by the application.
+The integrity and encryption keys are part of the TEE measurement.
+
+* hostfs is a directory residing within the insecure host that is mounted into the TEE.
+More hostfs directories can be mounted to a target directory by the application.
+There is no integrity checks done on this type of filesystem.
+
+The first dilemma with mounts is the source of the mount is untrusted for ext2 and hostfs.
+Only ext2 filesystem images support integrity and encryption where the integrity can be validated within the TEE.
+If the decryption key can be achieved securely within the TEE then the decryption can be done securely too without the key being passed in from the host.
+
+The second dilemma is that every host this runs on *can* have different disk layouts and the source of the mount may change every time.
+This is especially true in a container world where different filesystem images may be downloaded to different temporary locations that may differ with each run.
+
+When mounting file-systems into the TEE it is good to have the configuration that is secure being part fo the measured configuration of the application,
+but keep the host specific mount information separate and specifiable from separate configuration that is passed on the command-line.
+
+## TEE configuration
+
+A new node will be added to the JSON configuration that is used in the signing of the application through sign-sgx or package-sgx commands:
+
+```json
+mount: [
```

Comment:
TEE enforceable mount info (target, type, flags, etc, not source) might be included in secret release policy. With mount info represented as json array (following OCI spec),  it's complex to describe a policy such as "if only these 3 volumes  (with the specific properties) are mounted", or "as long as these 3 volumes are mounted". Intuitively, converting the array to name/value pairs, for example disk0:{target, type...}, disk1:{target, type ...}, might make the policy syntax simpler. But at this point, I don't have a full design. Starting with json array for mount info following OCI spec makes sense, but we should consider the possibility that it might need to change in the future.

Reply:
There is going to be a full re-design of the config at some point so I am expecting changes. 
The format you suggested was one of the designs I considered. I also thought about "target": {"type":"ext2", "flags","flag"} which I thought was a little cleaner, but it did not conform to the OCI spec and chose to not do that.
The parameters that are needed for ext2 that are used from the data mount parameter for keys we can also include in the flags as a "extencryptkey=<key>". Breaking with OCI would allow cleaner configuration though so these parameters can be individual parameters instead of flags.